#! /bin/sh /usr/share/dpatch/dpatch-run
## proxy_ftp_dirparse_crash.dpatch by Ulf Harnhammer
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Don't crash when FTP server sends back no spaces
## DP: http://issues.apache.org/bugzilla/show_bug.cgi?id=40733

@DPATCH@
diff -urNad apache2~/modules/proxy/mod_proxy_ftp.c apache2/modules/proxy/mod_proxy_ftp.c
--- apache2~/modules/proxy/mod_proxy_ftp.c	2006-07-12 05:38:44.000000000 +0200
+++ apache2/modules/proxy/mod_proxy_ftp.c	2007-06-07 01:18:13.000000000 +0200
@@ -517,6 +517,11 @@
             }
 
             filename = strrchr(ctx->buffer, ' ');
+
+	    if (filename == NULL) {
+	      ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, r->server,
+	      		"proxy: FTP: could not parse line %s", ctx->buffer);
+	    } else {
             *(filename++) = '\0';
 
             /* handle filenames with spaces in 'em */
@@ -543,6 +548,7 @@
                                    ap_escape_uri(p, filename),
                                    ap_escape_html(p, filename));
             }
+	    }
         }
         /* Try a fallback for listings in the format of "ls -s1" */
         else if (0 == ap_regexec(re, ctx->buffer, LS_REG_MATCH, re_result, 0)) {
